---
title: "Human COVID microglia scRNA low-level processing"
output: html_notebook
---
   
# Setup   
## Load packages etc.  
```{r}
library(tidyverse)
```
   
# Run pipelines   
## Set up environment   
   
Add to ~/.bashrc   
   
```{bash eval=FALSE}
# using pyenv in b1038
export PYENV_ROOT="/projects/b1038/tools/pyenv"
export PATH="$PYENV_ROOT/bin:$PATH"
eval "$(pyenv init --path)"
eval "$(pyenv init -)"
eval "$(pyenv virtualenv-init -)"

#for Nik's scRNA pipeline
source /projects/b1038/tools/snakemake/include.sh
```

Add snakemake config at ~/.slurm.json   
   
```{bash eval=FALSE}
{
    "__default__" : {
        "account" : "b1042",
        "partition" : "genomics",
        "mail_type" : "END,FAIL",
        "mail_user" : "rogangrant2022@u.northwestern.edu",
        "nodes" : "1"
    }
}
```

## Create environment   
### Install reticulate-friendly python   
```{bash eval=FALSE}
env PYTHON_CONFIGURE_OPTS="--enable-shared" pyenv install 3.9.5
```

### Requirements.in   
   
/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/requirements.in
   
```{python eval=FALSE}
snakemake==6.9.1
pandas==1.3.3
scvi-tools==0.14.3
scrublet==0.2.3
-e git+https://github.com/broadinstitute/CellBender.git@v0.2.0#egg=cellbender
torch-1.9.0+cu111-cp39-cp39-linux_x86_64.whl
plotnine==0.8.0
scanpy==1.8.2
```
   
### Compile   
```{bash eval=FALSE}
#pyenv global 3.9.5
cd /projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/
#python -mvenv human_microglia_venv
source /projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/human_microglia_venv/bin/activate
#pip install pip-tools
pip-compile
pip-sync
```
   
## Run in-house cellranger pipeline (my edits)   
### Submission script   
```{bash eval=FALSE}
cd /projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/merged_data/EBV1
screen
source /projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/human_microglia_venv/bin/activate
module load cellranger/6.1.2

snakemake -s /home/rag0151/utils/snakemake/pipelines/cellranger_pipeline_6_1_2 \
--cluster-config '/home/rag0151/.slurm.json' \
--config sample_csv_path='/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/raw/samplesheet_batch2.csv' \
data_dir='/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2' \
cellranger_dir='/software/cell-ranger/6.1.2/cellranger-6.1.2' \
transcriptome='/projects/b1038/Pulmonary/genomes/GRCh38_and_SARS-CoV-2_NC45512.2_and_EBV1_NC007605.1/' \
--cluster /projects/b1038/tools/snakemake/cluster.py \
-j 1000 \
-n

snakemake -s /home/rag0151/utils/snakemake/pipelines/cellranger_pipeline_6_1_2 \
--cluster-config '/home/rag0151/.slurm.json' \
--config sample_csv_path='/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/raw/samplesheet_batch2.csv' \
data_dir='/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2' \
cellranger_dir='/software/cell-ranger/6.1.2/cellranger-6.1.2' \
transcriptome='/projects/b1038/Pulmonary/genomes/GRCh38_and_SARS-CoV-2_NC45512.2_and_EBV1_NC007605.1/' \
--cluster /projects/b1038/tools/snakemake/cluster.py \
-j 1000
```
   
# Cell calling and ambient RNA removal (cellbender)   
## Setup   
### Snakefile   
   
~/aging_microglia_flu/220601_Snakefile_cellbender_batch2
   
```{python eval=FALSE}
EXPECTED_CELLS = {
  "SC294": 3088,
  "SC295": 9000
  }

rule all:
  input:
    expand("/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/cellbender/{sample}_filtered.h5", sample=["SC294", "SC295"])

rule cellbender:
  input:
    "/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/cellranger/{sample}/outs/raw_feature_bc_matrix.h5"
  output:
    "/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/cellbender/{sample}_filtered.h5"
  params:
    slurm__hours=6,
    slurm__cores=8,
    slurm__mem=40,
    slurm__partition="genomics-gpu",
    slurm__account="b1042",
    slurm__gres="gpu:a100:1",
    expected_cells = lambda wildcards, input: EXPECTED_CELLS.get(wildcards.sample)
  shell:
    """
    input=`realpath {input}`
    output=`realpath {output}`
    output="${{output%_filtered.h5}}.h5"
    cellbender remove-background \
    --cuda \
    --input "$input" \
    --output "$output" \
    --expected-cells {params.expected_cells} \
    --total-droplets-included 30000 \
    --fpr 0.01 \
    --epochs 150 \
    --learning-rate 5e-5
    """
```
   
## Run pipe   
```{bash eval=FALSE}
cd /projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/code/human_microglia_batch2
screen
source /projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/human_microglia_venv/bin/activate
snakemake -s ~/aging_microglia_flu/220601_Snakefile_cellbender_batch2 \
--cluster-config '/home/rag0151/.slurm.json' \
--cluster /projects/b1038/tools/snakemake/cluster.py \
-j 1000 \
-F \
-n

snakemake -s ~/aging_microglia_flu/220601_Snakefile_cellbender_batch2 \
--cluster-config '/home/rag0151/.slurm.json' \
--cluster /projects/b1038/tools/snakemake/cluster.py \
-j 1000 \
-F
```   
   
# Doublet calling (scrublet)      
## Setup   
### Snakefile   
   
~/aging_microglia_flu/220601_Snakefile_scrublet_batch2
   
```{python eval=FALSE}
import scrublet as scr
import scipy.io
import os
import gzip
import pandas as pd
import matplotlib.pyplot as plt
import subprocess

EXPECTED_DOUBLET_RATE = {
  "SC294": 0.016,
  "SC295": 0.061}
#Automatic detection is useless
CUTOFFS = {
  "SC294": 0.15,
  "SC295": 0.22}

rule all:
  input:
    expand("/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/scrublet/{sample}_doublets.csv", sample=["SC294", "SC295"]),
    expand("/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/scrublet/{sample}_threshold.txt", sample=["SC294", "SC295"]),
    expand("/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/scrublet/{sample}_histogram.pdf", sample=["SC294", "SC295"]),
    expand("/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/scrublet/{sample}_umap.pdf", sample=["SC294", "SC295"])

rule scrublet:
  input:
    "/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/cellranger/{sample}/outs/filtered_feature_bc_matrix/matrix.mtx.gz",
    "/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/cellranger/{sample}/outs/filtered_feature_bc_matrix/barcodes.tsv.gz"
  output:
    "/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/scrublet/{sample}_doublets.csv",
    "/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/scrublet/{sample}_threshold.txt",
    "/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/scrublet/{sample}_histogram.pdf",
    "/projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/data/human_microglia_batch2/scrublet/{sample}_umap.pdf"
    
  params:
    slurm__hours=1,
    slurm__cores=1,
    slurm__mem=8,
    edr = lambda wildcards, input: EXPECTED_DOUBLET_RATE.get(wildcards.sample),
    cutoff = lambda wildcards, input: CUTOFFS.get(wildcards.sample)
  run:
    subprocess.run("module load gcc/11.2.0", shell=True)
    counts_matrix = scipy.io.mmread(gzip.open(input[0])).T.tocsc()
    barcodes = pd.read_table(gzip.open(input[1]), header=None)
    scrub = scr.Scrublet(counts_matrix, expected_doublet_rate = params.edr)
    doublet_scores, doublets = scrub.scrub_doublets(min_counts=2,
                                                    min_cells=3,
                                                    min_gene_variability_pctl=85,
                                                    n_prin_comps=30)
    
    #just go with manual
    scrub.call_doublets(threshold = params.cutoff)
    
    #output doublet scores
    barcodes['doublet'] = doublet_scores
    barcodes.to_csv(output[0])
    
    #output cutoffs (not important now that we did manually)
    if not os.path.exists(output[1]):
        with open(output[1], 'w') as f:
            f.write(str(scrub.threshold_))
            
    #output theshold plot
    scrub.plot_histogram()
    plt.savefig(output[2])
    
    #output UMAP plot
    scrub.set_embedding('UMAP', scr.get_umap(scrub.manifold_obs_, n_neighbors=10, min_dist=0.2, metric='cosine'))
    scrub.plot_embedding('UMAP', order_points=True)
    plt.savefig(output[3])
```
   
## Run pipe   
```{bash eval=FALSE}
cd /projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq
screen
source /projects/b1038/Pulmonary/rgrant/microglia_aging_flu/scRNAseq/human_microglia_venv/bin/activate
module purge all
module load gcc/11.2.0 #required for annoy
snakemake -s ~/aging_microglia_flu/220601_Snakefile_scrublet_batch2 \
--cluster-config '/home/rag0151/.slurm.json' \
--cluster /projects/b1038/tools/snakemake/cluster.py \
-j 1000 \
-F \
-n

snakemake -s ~/aging_microglia_flu/220601_Snakefile_scrublet_batch2 \
--cluster-config '/home/rag0151/.slurm.json' \
--cluster /projects/b1038/tools/snakemake/cluster.py \
-j 1000 \
-F
```   
